src/timeout.ts

```diff
@@ -108,3 +108,165 @@ export class Timeout extends Promise<never> {
     );
   }
 }
+
+/** @internal */
+export type TimeoutContextOptions = LegacyTimeoutContextOptions | CSOTTimeoutContextOptions;
+
+/** @internal */
+export type LegacyTimeoutContextOptions = {
+  serverSelectionTimeoutMS: number;
+  waitQueueTimeoutMS: number;
+  socketTimeoutMS?: number;
+};
+
+/** @internal */
+export type CSOTTimeoutContextOptions = {
+  timeoutMS: number;
+  serverSelectionTimeoutMS: number;
+  socketTimeoutMS?: number;
+};
+
+function isLegacyTimeoutContextOptions(v: unknown): v is LegacyTimeoutContextOptions {
+  return (
+    v != null &&
+    typeof v === 'object' &&
+    'serverSelectionTimeoutMS' in v &&
+    typeof v.serverSelectionTimeoutMS === 'number' &&
+    'waitQueueTimeoutMS' in v &&
+    typeof v.waitQueueTimeoutMS === 'number'
+  );
+}
+
+function isCSOTTimeoutContextOptions(v: unknown): v is CSOTTimeoutContextOptions {
+  return (
+    v != null &&
+    typeof v === 'object' &&
+    'serverSelectionTimeoutMS' in v &&
+    typeof v.serverSelectionTimeoutMS === 'number' &&
+    'timeoutMS' in v &&
+    typeof v.timeoutMS === 'number'
+  );
+}
+
+/** @internal */
+export abstract class TimeoutContext {
+  static create(options: TimeoutContextOptions): TimeoutContext {
+    if (isCSOTTimeoutContextOptions(options)) return new CSOTTimeoutContext(options);
+    else if (isLegacyTimeoutContextOptions(options)) return new LegacyTimeoutContext(options);
+    else throw new MongoRuntimeError('Unrecognized options');
+  }
+
+  abstract get serverSelectionTimeout(): Timeout | null;
+
+  abstract get connectionCheckoutTimeout(): Timeout | null;
+
+  abstract get clearServerSelectionTimeout(): boolean;
+
+  abstract get clearConnectionCheckoutTimeout(): boolean;
+
+  abstract csotEnabled(): this is CSOTTimeoutContext;
+}
+
+/** @internal */
+export class CSOTTimeoutContext extends TimeoutContext {
+  timeoutMS: number;
+  serverSelectionTimeoutMS: number;
+  socketTimeoutMS?: number;
+
+  clearConnectionCheckoutTimeout: boolean;
+  clearServerSelectionTimeout: boolean;
+
+  private _maxTimeMS?: number;
+
+  private _serverSelectionTimeout?: Timeout | null;
+  private _connectionCheckoutTimeout?: Timeout | null;
+
+  constructor(options: CSOTTimeoutContextOptions) {
+    super();
+    this.timeoutMS = options.timeoutMS;
+
+    this.serverSelectionTimeoutMS = options.serverSelectionTimeoutMS;
+
+    this.socketTimeoutMS = options.socketTimeoutMS;
+
+    this.clearServerSelectionTimeout = false;
+    this.clearConnectionCheckoutTimeout = true;
+  }
+
+  get maxTimeMS(): number {
+    return this._maxTimeMS ?? -1;
+  }
+
+  set maxTimeMS(v: number) {
+    this._maxTimeMS = v;
+  }
+
+  csotEnabled(): this is CSOTTimeoutContext {
+    return true;
+  }
+
+  override get serverSelectionTimeout(): Timeout | null {
```

Comment:
The `override` keywords in the subclasses are not necessary as the parent class defines them as abstract.